packages/astro/src/core/build/static-build.ts

```diff
@@ -232,18 +276,39 @@ async function clientBuild(
 	return buildResult;
 }
 
-async function cleanSsrOutput(opts: StaticBuildOptions) {
-	const out = getOutDirWithinCwd(opts.settings.config.outDir);
+/**
+ * For each statically prerendered page, replace their SSR file with a noop.
+ * This allows us to run the SSR build only once, but still remove dependencies for statically rendered routes.
+ */
+async function cleanStaticOutput(opts: StaticBuildOptions, internals: BuildInternals) {
+	const allStaticFiles = new Set();
+	for (const pageData of eachPrerenderedPageData(internals)) {
+		allStaticFiles.add(internals.pageToBundleMap.get(pageData.moduleSpecifier));
+	}
+	const ssr = opts.settings.config.output === 'server';
+	const out = ssr ? opts.buildConfig.server : getOutDirWithinCwd(opts.settings.config.outDir);
 	// The SSR output is all .mjs files, the client output is not.
 	const files = await glob('**/*.mjs', {
 		cwd: fileURLToPath(out),
 	});
+
 	if (files.length) {
+		await eslexer.init;
 		// Remove all the SSR generated .mjs files
 		await Promise.all(
 			files.map(async (filename) => {
+				if (!allStaticFiles.has(filename)) {
+					return;
+				}
 				const url = new URL(filename, out);
-				await fs.promises.rm(url);
+				const text = await fs.promises.readFile(url, { encoding: 'utf8' });
```

Comment:
Can this be done in a plugin as well? Probably just have a `vite-plugin-prerender` plugin.

I think this can probably be done in the `generateBundles` hook and not need to read/write from the file after its been written.

Reply:
This can't be done in a plugin because it has to happen _after_ the SSR build runs. We are generating the `prerender` chunk, running it to statically generate the pages, then doing this step to remove the code here before the server is deployed.

This chunk will only contain the code pertaining to the `prerender` chunk, so it is safe to modify.

Reply:
Added some comments in [7aa1914](https://github.com/withastro/astro/pull/5297/commits/7aa191458d4ae2841cb19cac3b5abae3fe17bd3e)

Reply:
Are you saying that it has to happen after *generation*?

Reply:
I think I understand if that's the case.

Reply:
Yes, _after generation_ is the right way to phrase it.